ftl/common/builder.py

```diff
@@ -34,58 +50,129 @@ def __init__(self, ctx):
+    for runtime applications with dependencies.
+
+    It provides methods for generating appending layers and caching images
+    """
 
-    def BuildAppLayer(self, destination_path='srv'):
-        """Override."""
-        if destination_path is None:
-            destination_path = 'srv'
-        buf = cStringIO.StringIO()
-        logging.info('Starting to generate app layer tarfile from context...')
-        with tarfile.open(fileobj=buf, mode='w') as out:
-            for name in self._ctx.ListFiles():
-                content = self._ctx.GetFile(name)
-                info = tarfile.TarInfo(
-                    os.path.join(destination_path.strip("/"), name))
-                info.size = len(content)
-                out.addfile(info, fileobj=cStringIO.StringIO(content))
-        logging.info('Finished generating app layer tarfile from context.')
-
-        tar = buf.getvalue()
-        sha = 'sha256:' + hashlib.sha256(tar).hexdigest()
-
-        logging.info('Starting to gzip app layer tarfile...')
-        gzip_process = subprocess.Popen(
-            ['gzip', '-f'],
-            stdout=subprocess.PIPE,
-            stdin=subprocess.PIPE,
-            stderr=subprocess.PIPE)
-        gz = gzip_process.communicate(input=tar)[0]
-        logging.info('Finished gzipping tarfile.')
-        return gz, sha
+    def __init__(self, ctx, namespace, args, cache_version_str,
+                 descriptor_files):
+        super(RuntimeBase, self).__init__(ctx)
+        self._namespace = namespace
+        self._args = args
+        self._base_name = docker_name.Tag(self._args.base)
+        self._base_creds = docker_creds.DefaultKeychain.Resolve(
+            self._base_name)
+        self._target_image = docker_name.Tag(self._args.name)
+        self._target_creds = docker_creds.DefaultKeychain.Resolve(
+            self._target_image)
+        self._transport = transport_pool.Http(httplib2.Http, size=_THREADS)
+        self._cash = cache.Registry(
+            self._target_image.as_repository(),
+            self._target_creds,
+            self._transport,
+            cache_version=cache_version_str,
+            threads=_THREADS,
+            mount=[self._base_name])
+        self._base = docker_image.FromRegistry(
+            self._base_name, self._base_creds, self._transport)
+        self._base.__enter__()
+        self._descriptor_files = descriptor_files
+
+    def Build(self):
+        return
+
+    def AppendLayersIntoImage(self, lyr_imgs):
+        for i, lyr_img in enumerate(lyr_imgs):
+            if i == 0:
+                result_image = lyr_img
+                continue
+            img = lyr_img.GetImage()
+            diff_id = img.diff_ids()[0]
+            lyr = img.blob(img._diff_id_to_digest(diff_id))
+            overrides_dct = ftl_util.CfgDctToOverrides(
+                json.loads(img.config_file()))
+            if 'created' in overrides_dct:
```

Comment:
What's this for?

Reply:
`created` is the name of the key the Docker config_file for creation time but `creation_time` is the name of the value in the Overrides object.  Before I was using `creation_time` everywhere but this was not correct as then the config_file had the value `creation_time` which was not being used correctly (should have been `created`).  These lines are to generate an overrides with `creation_time` specified as it wasn't clear to me how to do it otherwise

Reply:
I put this code into CfgDctToOverrides, making it return an Overrides